src/rt/sched/behaviourcore.h

```diff
@@ -57,74 +57,308 @@ namespace verona::rt
+      ~(STATUS_SLOT_ACTIVE_FLAG | STATUS_NEXT_SLOT_READER_FLAG);
 
-    bool is_ready()
+    /**
+     * Points to the behaviour associated with this slot.
+     * This pointer is set only for reader slots.
+     * TODO: Change this to get behaviour address from start of allocation if
+     * snmalloc is used otherwise use this pointer.
+     */
+    std::atomic<BehaviourCore*> behaviour;
+
+  public:
+    Slot(Cown* __cown)
     {
-      return status.load(std::memory_order_acquire) == 1;
+      // Check that the last two bits are zero
+      assert(((uintptr_t)__cown & ~COWN_POINTER_MASK) == 0);
+      _cown.store((uintptr_t)__cown, std::memory_order_release);
+      status.store(0, std::memory_order_release);
+      behaviour.store(nullptr, std::memory_order_release);
     }
 
-    void set_move()
+    /**
+     * Returns true if the slot is acquired in read mode
+     */
+    bool is_read_only()
     {
-      status.store(1, std::memory_order_relaxed);
+      return (_cown.load(std::memory_order_acquire) & COWN_READER_FLAG) ==
+        COWN_READER_FLAG;
     }
 
-    void reset_status()
+    /**
+     * Mark the slot to be acquired in read mode
+     */
+    void set_read_only()
     {
-      status.store(0, std::memory_order_relaxed);
+      _cown.store(
+        _cown.load(std::memory_order_acquire) | COWN_READER_FLAG,
+        std::memory_order_release);
     }
 
-    void set_ready()
+    /**
+     * Returns true if the next slot wants to acquire in read mode
+     */
+    bool is_next_slot_read_only()
     {
-      status.store(1, std::memory_order_release);
+      return (
+        (status.load(std::memory_order_acquire) &
+         STATUS_NEXT_SLOT_READER_FLAG) == STATUS_NEXT_SLOT_READER_FLAG);
+    }
+
+    /**
+     * Returns true if all the slots in a behaviour haven't finished their
+     * acquire phase
+     */
+    bool is_wait_2pl()
+    {
+      return (_cown.load(std::memory_order_acquire) & COWN_2PL_READY_FLAG) !=
+        COWN_2PL_READY_FLAG;
     }
 
-    bool is_wait()
+    /**
+     * Mark the slot as ready to signify that its acquire phase is complete
+     */
+    void set_ready()
     {
-      return status.load(std::memory_order_relaxed) == 0;
+      _cown.store(
+        _cown.load(std::memory_order_acquire) | COWN_2PL_READY_FLAG,
+        std::memory_order_release);
     }
 
-    bool is_behaviour()
+    /**
+     * Mark the reader slot as active i.e. the behaviour is scheduled.
+     * Next reader in the queue can also be scheduled.
+     */
+    void set_active()
     {
-      return status.load(std::memory_order_relaxed) > 1;
+      assert(is_read_only());
+      while (true)
+      {
+        uintptr_t old_status_val = status.load(std::memory_order_acquire);
+        uintptr_t new_status_val = old_status_val | STATUS_SLOT_ACTIVE_FLAG;
+        if (status.compare_exchange_strong(
+              old_status_val, new_status_val, std::memory_order_acq_rel))
+          break;
+      }
     }
```

Comment:
```C++
    /**
     * Mark the reader slot as active i.e. the behaviour is scheduled.
     * Next reader in the queue can also be scheduled.
     */
    uintptr_t set_active_return_next()
    {
      assert(is_read_only());
      yield();
      return status.fetch_add(STATUS_SLOT_ACTIVE_FLAG, std::memory_order_acq_rel);
    }

   /**
     * Mark the reader slot as active i.e. the behaviour is scheduled.
     * Next reader in the queue can also be scheduled.
     */
    void set_active()
    {
      assert(is_read_only());
      assert(status == 0);
      yield();
      status.store(STATUS_SLOT_ACTIVE_FLAG, std::memory_order_release);
    }
```

Here `set_active` is used in the two cases where the 2PL has not completed, so there is no race.  And `set_active_get_next` returns the next value that was in the `status` before the call.